plugins/org.yakindu.sct.model.stext/src/org/yakindu/sct/model/stext/validation/STextNamesAreUniqueValidationHelper.java

```diff
@@ -0,0 +1,176 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.xtext.naming.QualifiedName;
+import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.service.OperationCanceledManager;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.validation.INamesAreUniqueValidationHelper;
+import org.eclipse.xtext.validation.NamesAreUniqueValidationHelper;
+import org.eclipse.xtext.validation.ValidationMessageAcceptor;
+
+/**
+ * @author rbeckmann
+ *
+ */
+public class STextNamesAreUniqueValidationHelper extends NamesAreUniqueValidationHelper implements INamesAreUniqueValidationHelper {
+	protected OperationCanceledManager operationCanceledManager = new OperationCanceledManager();
+
+	protected Map<QualifiedName, IEObjectDescription> nameMap;
+	
+	protected Map<EClass, Node<EClass>> superClasses = new HashMap<>();
+
+	@Override
+	public void checkUniqueNames(Iterable<IEObjectDescription> descriptions, ValidationMessageAcceptor acceptor) {
+		checkUniqueNames(descriptions, null, acceptor);
+	}
+
+	/**
+	 * <p>
+	 * {@inheritDoc}
+	 * </p>
+	 * The cancel indicator will be queried everytime a description has been
+	 * processed. It should provide a fast answer about its canceled state.
+	 */
+	@Override
+	public void checkUniqueNames(Iterable<IEObjectDescription> descriptions, CancelIndicator cancelIndicator,
+			ValidationMessageAcceptor acceptor) {
+		Iterator<IEObjectDescription> iter = descriptions.iterator();
+		this.nameMap = new HashMap<>();
+		if (!iter.hasNext())
+			return;
+		while (iter.hasNext()) {
+			IEObjectDescription description = iter.next();
+			checkDescriptionForDuplicatedName(description, acceptor);
+			operationCanceledManager.checkCanceled(cancelIndicator);
+		}
+	}
+
+	protected void checkDescriptionForDuplicatedName(IEObjectDescription description,
+			ValidationMessageAcceptor acceptor) {
+		QualifiedName qName = description.getName();
+		IEObjectDescription put = nameMap.put(qName, description);
+		if(put != null) {
+			EClass common = checkForCommonSuperClass(put, description, acceptor);
+			if(common != null) {
+				createDuplicateNameError(description, common, acceptor);
+				createDuplicateNameError(put, common, acceptor);
+			}
+		}
+	}
+	
+	protected Node<EClass> getSuperClassTree(EClass eClass) {
+		Node<EClass> node;
+		
+		node = superClasses.get(eClass);
+		
+		if(node == null) {
+			node = new Node<EClass>(eClass);
+			buildSuperClassTree(node);
+			superClasses.put(eClass, node);
+		}
+		
+		return node;
+	}
+
+	protected EClass checkForCommonSuperClass(IEObjectDescription one, IEObjectDescription two,
+			ValidationMessageAcceptor acceptor) {
+		
+		Node<EClass> superClassesOne = getSuperClassTree(one.getEClass());
+		Node<EClass> superClassesTwo = getSuperClassTree(two.getEClass());
+		
+		List<EClass> flatOne = superClassesOne.flatten();

```

Comment:
If you use only the flat representation of the super type hierachy there is no need to build a graph data structure (Node<T>).